### DISCUSSION TITLE
[Released] Expose the commit sha used for a particular branch deployment

### DISCUSSION THREAD

**Discussion by Js-Brecht** [vercel/vercel#6087] [Ideas] [answered]
I am operating a monorepo, and recently started using Vercel.  I am trying to work out how to get the "Ignore Build Step" to work accurately when I may have a collection of commits being made to multiple projects, and the _latest_ may not be a change to the project being deployed, but there _**is**_ a commit that was a change to the project and _**should**_ be deployed.

I need a frame of reference for the latest commit that was built.  With that, I could look at what latest commit changed the project in question, compare it to the latest commit built, and if they differ than I build & deploy.

My git branches are are structured like this:

* `master`
  * All pull requests are based off of `master`, so they're getting staged here and deployed later.
  * I have many projects that get staged in master, and many of them have their own deployment mechanisms... but they all use this same branch deployment structure
  * Changes to the Vercel project _**will**_ trigger a deployment here, because all pull requests get merged into `master` as squash commits.
  * These preview deployments are useful because they act as a sort of temporary staging environment that can be used to test all changes that are about to get deployed to other environments
* `demo`
  * This branch should trigger a "demo" deployment, which gets its own domain name & environment variables.
  * This deployment uses mocked data & api calls, and is used for demoing the product to clients.
  * I prefix environment variables for this branch with `DEMO_`, and reduce them during the build to favor the ones with the prefix.
* `staging`
  * Mostly like the `master` branch, except this is a permanent staging environment with its own domain name.  This uses api calls against our testing database/backend.
* `production`
  * This branch should trigger a "production" deployment.  Same basic design as the `demo` branch, but uses production values, domain name, and api

When I merge master into `staging`, `demo`, or `production`, I fast-forward only so the branches stay even.  The problem I have is when I get multiple pull requests merged into `master`, the latest commit that winds up in `staging` / `production` are sometimes _**not**_ the commit that changed the project that needs to be deployed.

I'm looking for something like `$VERCEL_LAST_BUILD_SHA`.  I'd expect its value to be the last commit that was built for whatever branch that's being deployed.  Then I could do something like this:

```sh
change_commit=$(git log -n 1 --pretty=%H -- ./);
if [[ "$change_commit" == "$VERCEL_LAST_BUILD_SHA" ]]; then
  exit 0;
else
  exit 1;
fi
```

**Comment by amyegan** [comment, marked as answer]
Thanks for the suggestion! This is available now 😄 

> A new system environment variable: [VERCEL_GIT_PREVIOUS_SHA](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables) is now available in the [Ignored Build Step](https://vercel.com/docs/concepts/projects/overview#ignored-build-step), allowing scripts to compare changes against the SHA of the last successful deployment for the current project, and branch.
— [Improvements and Fixes](https://vercel.com/changelog/september-2022-papercuts)
